apps/stable_diffusion/web/ui/h2ogpt.py

```diff
@@ -168,10 +172,106 @@ def chat(curr_system_message, history, device, precision):
         history[-1][1] = partial_text["response"]
         yield history
 
-    return history
 
+userpath_selector = gr.Textbox(
+    label="Document Directory",
+    value=str(os.path.abspath("apps/language_models/langchain/user_path/")),
+    interactive=True,
+    container=True,
+)
+
+with gr.Blocks(title="DocuChat Upload") as h2ogpt_upload:
+    import pathlib
+
+    upload_path = None
+    database = None
+    database_directory = os.path.abspath(
+        "apps/language_models/langchain/db_path/"
+    )
+
+    def read_path():
+        global upload_path
+        filenames = [
+            [f]
+            for f in os.listdir(upload_path)
+            if os.path.isfile(os.path.join(upload_path, f))
+        ]
+        filenames.sort()
+        return filenames
+
+    def upload_file(f):
+        names = []
+        for tmpfile in f:
+            name = tmpfile.name.split("/")[-1]
+            basename = os.path.join(upload_path, name)
+            with open(basename, "wb") as w:
+                with open(tmpfile.name, "rb") as r:
+                    w.write(r.read())
+        update_or_create_db()
+        return read_path()
+
+    def update_userpath(newpath):
+        global upload_path
+        upload_path = newpath
+        pathlib.Path(upload_path).mkdir(parents=True, exist_ok=True)
+        return read_path()
+
+    def update_or_create_db():
+        global database
+        global upload_path
+
+        sources = path_to_docs(
+            upload_path,
+            verbose=True,
+            fail_any_exception=False,
+            n_jobs=-1,
+            chunk=True,
+            chunk_size=512,
+            url=None,
+            enable_captions=False,
+            captions_model=None,
+            caption_loader=None,
+            enable_ocr=False,
+        )
 
-with gr.Blocks(title="H2OGPT") as h2ogpt_web:
+        pathlib.Path(database_directory).mkdir(parents=True, exist_ok=True)
+
+        database = create_or_update_db(
+            "chroma",
+            database_directory,
+            "UserData",
+            sources,
+            False,
+            True,
+            True,
+            "sentence-transformers/all-MiniLM-L6-v2",
+        )
+
+    update_userpath(
+        os.path.abspath("apps/language_models/langchain/user_path/")
+    )
+    update_or_create_db()
+
+    with gr.Column():
+        text = gr.DataFrame(
+            col_count=(1, "fixed"),
+            type="array",
+            label="Documents",
+            value=read_path(),
+        )
+        with gr.Row():
+            upload = gr.UploadButton(
+                label="Upload documents",
+                file_count="multiple",
+            )
+            upload.upload(fn=upload_file, inputs=upload, outputs=text)
+            userpath_selector.render()
+            userpath_selector.input(
+                fn=update_userpath, inputs=userpath_selector, outputs=text
+            ).then(fn=update_or_create_db)
```

Comment:
I think you should add this code to a different file. Because on using both the features in the same session, the database creation happens to be running twice.

Reply:
I don't see it running twice. If you mean that it runs on every file upload and once at the start that's expected behavior (in case the user populates the directory before the web UI is started).

Reply:
Could you also verify the amount of time both ways? With a document sized in MBs?

Reply:
After adding those files and starting up it takes 110 seconds. I've also verified that it's only running once per upload and once at the start.

Reply:
Just a small question, why do I see this
```
0it [00:00, ?it/s][ProgressParallel(n_jobs=-1)]: Using backend MultiprocessingBackend with 12 concurrent workers.
Ingesting file: /home/vivek/work/shark-vivekkhandelwal1/apps/language_models/langchain/user_path/./Abraham Silberschatz, Greg Gagne, Peter B. Galvin - Operating System Concepts-Wiley (2018).pdf
Ingesting file: /home/vivek/work/shark-vivekkhandelwal1/apps/language_models/langchain/user_path/./operating_system.txt
 50%|██████████████████████████████████████████████████████████████████████████████████████████████                                                                                              | 1/2 [00:00<00:00,  1.99it/s][ProgressParallel(n_jobs=-1)]: Batch computation too fast (0.02492046356201172s.) Setting batch_size=2.
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:22<00:00, 11.27s/it]
0it [00:00, ?it/s]
Loading and updating db
Found 0 new sources (0 have no hash in original source, so have to reprocess for migration to sources with hash)
Removing 0 duplicate files from db because ingesting those as new documents
```
Even before choosing/running any model?

Reply:
Shouldn't all this run when we select DocuChat? This way, those who don't want to run DocuChat would also have to go through this.

Reply:
I see what you mean, though I imagine someone not running DocuChat wouldn't be populating the directory.

Reply:
Okay, I've changed it to only run upon switching to either DocuChat tabs for the first time.